mllib/src/main/scala/org/apache/spark/ml/regression/FactorizationMachines.scala

```diff
@@ -0,0 +1,757 @@
+
+  /**
+   * Set the initial step size for the first step (like learning rate).
+   * Default is 1.0.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setStepSize(value: Double): this.type = set(stepSize, value)
+  setDefault(stepSize -> 1.0)
+
+  /**
+   * Set the convergence tolerance of iterations.
+   * Default is 1E-6.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setTol(value: Double): this.type = set(tol, value)
+  setDefault(tol -> 1E-6)
+
+  /**
+   * Set the solver algorithm used for optimization.
+   * Default is adamW.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setSolver(value: String): this.type = set(solver, value)
+  setDefault(solver -> AdamW)
+
+  /**
+   * Sets the value of param [[loss]].
+   * - "logisticLoss" is used to classification, label must be in {0, 1}.
+   * - "squaredError" is used to regression.
+   * Default is logisticLoss.
+   *
+   * @group setParam
+   */
+  @Since("3.0.0")
+  def setLoss(value: String): this.type = set(loss, value)
+  setDefault(loss -> LogisticLoss)
+
+  override protected[spark] def train(dataset: Dataset[_]): FactorizationMachinesModel = {
+    val handlePersistence = dataset.rdd.getStorageLevel == StorageLevel.NONE
+    train(dataset, handlePersistence)
+  }
+
+  protected[spark] def train(
+      dataset: Dataset[_],
+      handlePersistence: Boolean): FactorizationMachinesModel = instrumented { instr =>
+    val instances: RDD[OldLabeledPoint] =
+      dataset.select(col($(labelCol)), col($(featuresCol))).rdd.map {
+        case Row(label: Double, features: Vector) =>
+          OldLabeledPoint(label, features)
+      }
+
+    if (handlePersistence) instances.persist(StorageLevel.MEMORY_AND_DISK)
+
+    instr.logPipelineStage(this)
+    instr.logDataset(dataset)
+    instr.logParams(this, numFactors, fitBias, fitLinear, regParam,
+      miniBatchFraction, initStd, maxIter, stepSize, tol, solver, loss)
+
+    val numFeatures = instances.first().features.size
+    instr.logNumFeatures(numFeatures)
+
+    // initialize coefficients
+    val coefficientsSize = $(numFactors) * numFeatures +
+      (if ($(fitLinear)) numFeatures else 0) +
+      (if ($(fitBias)) 1 else 0)
+    val initialCoefficients =
+      Vectors.dense(Array.fill($(numFactors) * numFeatures)(Random.nextGaussian() * $(initStd)) ++
+        (if ($(fitLinear)) Array.fill(numFeatures)(0.0) else Array.empty[Double]) ++
+        (if ($(fitBias)) Array.fill(1)(0.0) else Array.empty[Double]))
+
+    val data = instances.map { case OldLabeledPoint(label, features) => (label, features) }
+
+    // optimize coefficients with gradient descent
+    val gradient = BaseFactorizationMachinesGradient.parseLoss(
+      $(loss), $(numFactors), $(fitBias), $(fitLinear), numFeatures)
+
+    val updater = $(solver) match {
+      case GD => new SquaredL2Updater()
+      case AdamW => new AdamWUpdater(coefficientsSize)
+    }
+
+    val optimizer = new GradientDescent(gradient, updater)
```

Comment:
BTW, If both FM and MLP still need mini-batch solver, we may move it to the .ml side in the future, to avoid vector conversion.

Reply:
> BTW, If both FM and MLP still need mini-batch solver, we may move it to the .ml side in the future, to avoid vector conversion.

I agree, and I think GradientDescent supports aggregateDepth parameter, seed parameter, weightCol parameter and logInfo loss value per epoch will be better.